### PR TITLE
OCM-3704 | fix: Update error message for create/describe/list/delete machine pool upgrade policy

### PR DESCRIPTION
**Pull request by Mischulee**
https://issues.redhat.com/browse/OCM-3704

### CODE REVIEW

**Comment by Mischulee** [comment]
Examples:

 ./rosa upgrade machinepool  -c ying-up1
Before:
Error: expected exactly one command line parameter containing the id of the machine pool
Usage:
  rosa upgrade machinepool [flags]
...

Failed to execute root command: expected exactly one command line parameter containing the id of the machine pool



After:
Error: machine pool id is missing
Usage:
  rosa upgrade machinepool [flags]
...

Failed to execute root command: machine pool id is missing


======================================================================================
 ./rosa list upgrade machinepool  -c mparilov
Before:
Error: unknown command "mparilov-hcp2" for "rosa list upgrades"
Usage:
  rosa list upgrades [flags]
...

Failed to execute root command: unknown command "mparilov-hcp2" for "rosa list upgrades"



After:
Error: machine pool id is missing
Usage:
  rosa list upgrades [flags]
 ...

Failed to execute root command: machine pool id is missing

File: cmd/describe/upgrade/cmd.go [hunk 1]
```diff
@@ -38,7 +39,15 @@ var Cmd = &cobra.Command{
   rosa describe upgrade`,
 	Run:    run,
 	Hidden: false,
-	Args:   cobra.NoArgs,
+	Args: func(cmd *cobra.Command, argv []string) error {
```
Comment (robpblake) [inline cmd/describe/upgrade/cmd.go:42]: @Mischulee We are using the same function in multiple locations. Could you please refactor into a method that returns the function to be used. For example:

```
Args: NewUpgradeArgsFunction(),
```

where `NewUpgradeArgsFunction()` is a function that returns the `Args` func e.g:

```
func NewUpgradeArgsFunction() func(cmd *cobra.Command, argv []string) {
 return {
        flags := cmd.Flags()
	machinepoolID, _ := flags.GetString("machinepool")

	if !machinepool.MachinePoolKeyRE.MatchString(machinepoolID) {
		return fmt.Errorf("expected a valid identifier for the machine pool")
	}
	return nil
  }
}
```

Could you also the please add unit tests for the function to make sure it behaves as we expect. I'd suggest creating a new package in `pkg`, perhaps `upgrade` and moving the function and tests there.

There are lots of examples of how we use Ginkgo to write unit tests in the project. One that provides a good example of testing the flags on a `cobra.Command` is `pkg/output/flag_test.go`. Take a look and see if you are able to write your own test. If you have any other questions or need some pointers, please just connect with me on Slack.

cheers,

Rob

File: cmd/upgrade/machinepool/cmd.go [hunk 2]
```diff
@@ -53,7 +53,7 @@ var Cmd = &cobra.Command{
 	Args: func(_ *cobra.Command, argv []string) error {
 		if len(argv) != 1 {
 			return fmt.Errorf(
-				"expected exactly one command line parameter containing the id of the machine pool",
+				"machine pool id is missing",
```
Comment (robpblake) [inline cmd/upgrade/machinepool/cmd.go:56]: @Mischulee Is there a reason why the message is different and we are not verifying that the machine pool id matches the regex as we do in the other functions?

**Review by robpblake — CHANGES_REQUESTED** [review]
@Mischulee I've left a comment about refactoring to make the code here re-usable and also adding some unit tests.

File: cmd/upgrade/machinepool/cmd.go (hunk 2 above)
Comment (Mischulee) [reply cmd/upgrade/machinepool/cmd.go:56]: In this case, for example this:.  ./rosa upgrade machinepool -c mparilov
The machinepool is not flag as in other cases, so we can not use the same function. But I can add additional check for regex.

File: cmd/describe/upgrade/cmd.go (hunk 1 above)
Comment (Mischulee) [reply cmd/describe/upgrade/cmd.go]: @robpblake Thank you for a helpful review! I refactored the function and added unit test as suggested.

File: pkg/upgrade/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	"github.com/openshift/rosa/pkg/machinepool"
+)
+
+func NewUpgradeArgsFunction() func(cmd *cobra.Command, argv []string) error {
+	return func(cmd *cobra.Command, argv []string) error {
+		flags := cmd.Flags()
+		machinepoolID, _ := flags.GetString("machinepool")
+
+		if !machinepool.MachinePoolKeyRE.MatchString(machinepoolID) {
+			return fmt.Errorf("expected a valid identifier for the machine pool")
```
Comment (robpblake) [inline pkg/upgrade/upgrade.go:17]: @Mischulee We have this string repeated 3 times in this PR, so I'd suggest to make it a `const` in the `upgrade` package.

File: pkg/upgrade/upgrade.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/rosa/pkg/machinepool"
+)
+
+func NewUpgradeArgsFunction() func(cmd *cobra.Command, argv []string) error {
+	return func(cmd *cobra.Command, argv []string) error {
+		flags := cmd.Flags()
+		machinepoolID, _ := flags.GetString("machinepool")
```
Comment (robpblake) [inline pkg/upgrade/upgrade.go:14]: @Mischulee We shouldn't be ignoring the `error` here. Could you please add a check for the `err != nil` and return it from the function.

File: pkg/upgrade/upgrade_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/spf13/cobra"
+)
+
+func TestReporter(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "reporter testing")
```
Comment (robpblake) [inline pkg/upgrade/upgrade_test.go:13]: @Mischulee Could you update the text here to correctly identify the test suite. Thanks!

File: pkg/upgrade/upgrade_test.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package upgrade
```
Comment (robpblake) [inline pkg/upgrade/upgrade_test.go:1]: @Mischulee Great unit test. Really easy to read!

File: pkg/upgrade/upgrade.go (hunk 3 above)
Comment (robpblake) [reply pkg/upgrade/upgrade.go:17]: Alternatively you could make the `error` itself a package-level var and return that e.g:

```
var (
	InvalidMachinePoolIdentifier = fmt.Errorf("expected a valid identifier for the machine pool")
)
```

```
if !machinepool.MachinePoolKeyRE.MatchString(machinepoolID) {
			return InvalidMachinePoolIdentifier
		}
```

File: cmd/upgrade/machinepool/cmd.go [hunk 7]
```diff
@@ -53,9 +54,14 @@ var Cmd = &cobra.Command{
 	Args: func(_ *cobra.Command, argv []string) error {
```
Comment (robpblake) [inline cmd/upgrade/machinepool/cmd.go:54]: @Mischulee How about you make the source of the `machinepool` name a `bool` on your `NewUpgradeArgsFunction`?  So something like:

```
func NewUpgradeArgsFunction(flagSource bool) func(cmd *cobra.Command, argv []string) error {
	return func(cmd *cobra.Command, argv []string) error {
		var machinepoolID string
		var err error

		if flagSource {
			flags := cmd.Flags()
			if machinepoolID, err = flags.GetString(machinePoolFlagName); err != nil {
				return err
			}
		} else {
			if len(argv) != 1 {
				return InvalidMachinePoolIdentifier
			}
			machinepoolID = argv[0]
		}

		if !machinepool.MachinePoolKeyRE.MatchString(machinepoolID) {
			return InvalidMachinePoolIdentifier
		}
		return nil
	}
}
```

That means we have a single function for all scenarios, and you can add a few more unit tests to verify the flow via the args as well as via the flag.

File: pkg/upgrade/upgrade.go (hunk 3 above)
Comment (Mischulee) [reply pkg/upgrade/upgrade.go]: You are right, I added the second option.

File: pkg/upgrade/upgrade.go (hunk 4 above)
Comment (Mischulee) [reply pkg/upgrade/upgrade.go]: Fixed.

File: pkg/upgrade/upgrade_test.go (hunk 5 above)
Comment (Mischulee) [reply pkg/upgrade/upgrade_test.go]: Sorry I havent noticed, fixed :sweat_smile:

File: cmd/upgrade/machinepool/cmd.go (hunk 7 above)
Comment (Mischulee) [reply cmd/upgrade/machinepool/cmd.go]: You are right, this looks good and now it will be more reusable. I added more unit tests for this.

**Comment by robpblake** [comment]
Nice work @Mischulee 🚀

**Pull request merged**